modules/andes-core/broker/src/main/java/org/dna/mqtt/moquette/messaging/spi/impl/ProtocolProcessor.java

```diff
@@ -821,6 +824,46 @@ void processSubscribe(ServerChannel session, SubscribeMessage msg, String client
         } else {
             log.error("Not sending sub ack message since client " + clientID + " does not have permission to subscribe to all given subscriptions.");
         }
+
+        // Added by WSO2.
+        // Retain messages will handled inside ProtocolProcessor.
+        for (SubscribeMessage.Couple req : msg.subscriptions()) {
+            sendRetainMessageToSubscriber(req.getTopicFilter(), clientID);
+        }
+    }
+
+    /**
+     * Sends retain messages if exist upon subscription.
+     * Method written by WSO2.
+     *
+     * @param topicName topic name of the subscription.
+     * @param clientID subscription client id
+     */
+    private void sendRetainMessageToSubscriber(String topicName, String clientID) {
+
+        // Send retained message if available
+        try {
+
+            List<DeliverableAndesMetadata> metadataList = MessagingEngine.getInstance().getRetainedMessageByTopic(topicName);
+
+            for (DeliverableAndesMetadata metadata : metadataList) {
+                AndesContent content = MessagingEngine.getInstance().getRetainedMessageContent(metadata);
```

Comment:
implement required methods in Andes and invoke them there... 
Andes is the facade for Andes-core. Not messaging engine.